### PR TITLE
Refine Contract unittest

### PR DESCRIPTION
**Pull request by Jasmin3q**
# Description

Refine unittest for promptflow.contracts.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/contracts/flow.py [hunk 1]
```diff
@@ -583,7 +583,10 @@ def is_normal_node(self, node_name):
 
     def is_llm_node(self, node):
         """Given a node, return whether it uses LLM tool."""
-        return node.type == ToolType.LLM
+        if isinstance(node, Node):
+            return node.type == ToolType.LLM
+        else:
+            raise TypeError(f"Unexpected node type {type(node)}")
```
Comment (PeiwenGaoMS) [inline src/promptflow/promptflow/contracts/flow.py:589]: Why change this? Is there a possible reason why the node is not of type `Node`?

File: src/promptflow/tests/executor/unittests/contracts/test_run_info.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import pytest
+from datetime import datetime
+from promptflow.contracts.run_info import Status, RunInfo, FlowRunInfo
+
+
+@pytest.mark.unittest
+def test_status_is_terminated():
```
Comment (PeiwenGaoMS) [inline src/promptflow/tests/executor/unittests/contracts/test_run_info.py:7]: You can use `@pytest.mark.parametrize`, maybe like:
```python
@pytest.mark.parametrize(
  "status, is_terminated",
  [
    (Status.Completed, True),
    ...
  ]
)
def test_status_is_terminated(status, is_terminated):
  assert Status.is_terminated(status) == is_terminated
```

File: src/promptflow/promptflow/contracts/flow.py (hunk 1 above)
Comment (Jasmin3q) [reply src/promptflow/promptflow/contracts/flow.py:589]: The GPT generated test code remind me of this corner case. If it is not necessary, I'll remove the change in the next commit.

File: src/promptflow/tests/executor/unittests/contracts/test_tool.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+
+class TestStatus(Enum):
+    Running = 1
+    Preparing = 2
+    Completed = 3
+
+
+@pytest.mark.unittest
+def test_deserialize_enum():
```
Comment (PeiwenGaoMS) [inline src/promptflow/tests/executor/unittests/contracts/test_tool.py:18]: Can use `@pytest.mark.parametrize` too.

File: src/promptflow/promptflow/contracts/flow.py (hunk 1 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/contracts/flow.py:589]: I don't think it is necessary but not sure, you can double confirm with @thy09.

**Review by guming-llc — DISMISSED** [review]

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py [hunk 4]
```diff
@@ -58,3 +71,373 @@ def test_node_condition_conflict(self):
...
+        assert flow.nodes[0].use_variants is False
+        assert flow.nodes[0].inputs == flow.node_variants["print_val"].variants["variant1"].node.inputs
+        # test when node.use_variants is False
+        flow = Flow.from_yaml(flow_yaml)
+        flow.nodes[0].use_variants = False
+        tmp_nodes = flow.nodes
+        flow._apply_default_node_variants()
+        assert flow.nodes == tmp_nodes
+
+    def test_apply_default_node_variant(self):
```
Comment (zzn2) [inline src/promptflow/tests/executor/unittests/contracts/test_flow.py:135]: Most of the cases are not unit test style. Unit tests should test only one method, and cover all the conditions (branches). 
The current code is most likely to be an integration test. It depends on external methods/data.

Let's discuss offline for the details tomorrow.

File: src/promptflow/promptflow/contracts/tool.py [hunk 5]
```diff
@@ -254,7 +256,8 @@ def serialize(self) -> dict:
         """
 
         data = {}
-        data["type"] = ([t.value for t in self.type],)
+        # Delete the brackets outside data["type"] according to deserialize function
```
[the remark this answers is not in the document]
Comment (Jasmin3q) [reply src/promptflow/promptflow/contracts/tool.py:259]: Deleted.

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py [hunk 6]
```diff
@@ -1,22 +1,35 @@
...
 
 
+# Is this unittest or e2etest?
 @pytest.mark.e2etest
 class TestFlowContract:
     @pytest.mark.parametrize(
         "flow_folder, expected_connection_names",
         [
             ("web_classification", {"azure_open_ai_connection"}),
+            ("basic-with-connection", {"azure_open_ai_connection"}),
```
[the remark this answers is not in the document]
Comment (Jasmin3q) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py:35]: This is an existing flow in test_configs, should I change the name for it? It may affect other test cases.

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py (hunk 4 above)
Comment (Jasmin3q) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py:202]: For some functions must depend on external data such as get_connection_names, I choose to retain the e2etest in the current file. The other test cases are now changed to simple examples.

**Review by guming-llc — APPROVED** [review]

File: src/promptflow/promptflow/contracts/tool.py [hunk 7]
```diff
@@ -254,7 +255,7 @@ def serialize(self) -> dict:
         """
 
         data = {}
-        data["type"] = ([t.value for t in self.type],)
+        data["type"] = [t.value for t in self.type]
```
Comment (zzn2) [inline src/promptflow/promptflow/contracts/tool.py:258]: Why updating this, is it by design?

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zzn2) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py:34]: Let's keep it as-is. If we want to change the name, do it in a separate PR later.

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py [hunk 8]
```diff
@@ -58,3 +74,536 @@ def test_node_condition_conflict(self):
...
+        tool2 = Tool(name="tool2", type=ToolType.PYTHON, inputs={}, module="module")
+        node1 = Node(name="node1", tool="tool1", inputs={}, module="yaml")
+        node2 = Node(name="node2", tool="tool2", inputs={}, module="module")
+        tools = [tool1, tool2]
+        nodes = [node1, node2]
+
+        with pytest.raises(FailedToImportModule) as e:
+            Flow._import_requisites(tools, nodes)
+
+        assert str(e.value).startswith("Failed to import modules with error:")
```
Comment (zzn2) [inline src/promptflow/tests/executor/unittests/contracts/test_flow.py:163]: We can check more here.
* If failed in tool importing, the error message looks like "Import tool ... module ... failed."
* If failed in node importing, the error message looks like "Import node ... provider module ... failed."

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py [hunk 9]
```diff
@@ -58,3 +74,536 @@ def test_node_condition_conflict(self):
...
+        )
+        # test when node_variants is None
+        assert Flow._apply_default_node_variant(node, None) == node
+
+        # test when node.name is not in node_variants
+        node_variants1 = {"test": NodeVariants(default_variant_id="variant1", variants={"variant1": node_variant})}
+        assert Flow._apply_default_node_variant(node, node_variants1) == node
+        node_variants2 = {"print_val": NodeVariants(default_variant_id="test", variants={"variant1": node_variant})}
+        # test when default_variant_id is not in variants
+        assert Flow._apply_default_node_variant(node, node_variants2) == node
```
Comment (zzn2) [inline src/promptflow/tests/executor/unittests/contracts/test_flow.py:219]: Can we do these tests with ` @pytest.mark.parametrize`?

File: src/promptflow/tests/executor/unittests/contracts/test_multimedia.py [hunk 10]
```diff
@@ -20,4 +20,12 @@ def test_constructors(self):
             assert b._mime_type == "image/*"
             assert b._hash == "a94a8fe5"
             assert b.to_base64() == "dGVzdA=="
+            assert b.to_base64(True) == "data:image/*;base64,dGVzdA=="
+            assert b.to_base64(True, True) == {"data:image/*;base64": "dGVzdA=="}
```
Comment (zzn2) [inline src/promptflow/tests/executor/unittests/contracts/test_multimedia.py:24]: Can we call these with argument names? Otherwise, we can't tell what each `True` stands for.
```suggestion
            assert b.to_base64(with_type=True, dict_type=True) == {"data:image/*;base64": "dGVzdA=="}
```

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py (hunk 4 above)
Comment (zzn2) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py:205]: Thanks. Sounds good to me.

File: src/promptflow/promptflow/contracts/tool.py (hunk 7 above)
Comment (Jasmin3q) [reply src/promptflow/promptflow/contracts/tool.py:258]: This data type is not compatible with the deserialize function. According to the logic in deserialize function, there should not be brackets outside [t.value for t in self.type].

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py (hunk 9 above)
Comment (Jasmin3q) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py]: Added parametrize but the pre-commit makes it take up a lot of lines.

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py (hunk 8 above)
Comment (Jasmin3q) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py]: Added check for tool and node import error.

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py [hunk 11]
```diff
@@ -1,22 +1,35 @@
...
+)
+from promptflow.contracts.tool import ConnectionType, InputDefinition, Tool, ToolType, ValueType
+from promptflow.exceptions import UserErrorException
 
 from ...utils import WRONG_FLOW_ROOT, get_flow_package_tool_definition, get_yaml_file
 
 PACKAGE_TOOL_BASE = Path(__file__).parent.parent.parent / "package_tools"
 
 
+# Is this unittest or e2etest?
```
[the remark this answers is not in the document]
Comment (Jasmin3q) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py]: Keep these existing tests as e2etests because they refer external datas, and the new added tests will be marked as unittest.

**Review by guming-llc — APPROVED** [review]

File: src/promptflow/tests/executor/unittests/contracts/test_flow.py (hunk 9 above)
Comment (zzn2) [reply src/promptflow/tests/executor/unittests/contracts/test_flow.py]: It's fine. I like the new style.

**Review by zzn2 — APPROVED** [review]
